rllib/policy/eager_tf_policy.py

```diff
@@ -331,46 +332,59 @@ def compute_actions(self,
                         prev_reward_batch),
                 })
 
-            # Custom sampler fn given (which may handle self.exploration).
-            if action_sampler_fn is not None:
-                state_out = []
-                action, logp = action_sampler_fn(
-                    self,
-                    self.model,
-                    input_dict,
-                    self.observation_space,
-                    self.action_space,
-                    explore,
-                    self.config,
-                    timestep=timestep)
             # Use Exploration object.
-            else:
-                with tf.variable_creator_scope(_disallow_var_creation):
-                    # Call the exploration before_compute_actions hook.
-                    self.exploration.before_compute_actions(timestep=timestep)
-
-                    model_out, state_out = self.model(input_dict,
-                                                      state_batches, seq_lens)
-                    action, logp = self.exploration.get_exploration_action(
-                        model_out,
-                        self.dist_class,
+            with tf.variable_creator_scope(_disallow_var_creation):
+                if action_sampler_fn:
+                    dist_class = dist_inputs = None
+                    state_out = []
+                    actions, logp = self.action_sampler_fn(
+                        self,
                         self.model,
+                        input_dict[SampleBatch.CUR_OBS],
+                        explore=explore,
+                        timestep=timestep)
+                else:
+                    # Exploration hook before each forward pass.
+                    self.exploration.before_compute_actions(
```

Comment:
Shouldn't the hook be called in all cases including action_sampler_fn?

Reply:
Not sure. I was thinking of `action_sampler_fn` as a completely custom way of doing things. So the user would have to apply exploration him/herself.
That being said: No one is using `action_sampler_fn` right now, anyway, so I guess it doesn't matter much.